Property based testing with StreamData

SophieDeBenedetto opened the pull request:
Extracted vote calculation logic into its own module and tested with StreamData

@michaelstalker my question for you is: how much property testing + stream data content do you think we need for the workshop on day 2? How many such tests do we want to see participants write? I'm thinking probably more than what I have here? If so, what other opportunities for property-based testing can we find in our app--more cases for vote calculation and/or other features to test?

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,68 @@
+        cond do
+          is_list(winner) ->
+            winner
+            |> Enum.max()
```
Did you mean to leave lines 44-45 like this? We're not storing the results of the pipe operation.

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,68 @@
+        if is_list(winner) do
+          length = winner
+          |> length
+          assert length == 2
```
Lines 57-59 would be a little simpler if they were written like this:
```ex
assert length(winner) == 2
```

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,68 @@
+
+          sorted = winner
+          |> Enum.sort()
+          assert sorted = winner
```
We could consolidate lines 61-63 like this:
```ex
assert Enum.sort(winner) == winner
```

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,68 @@
+      [user: user]
+    end
+
+    test "winning value is a list or a integer", %{user: user} do
```
I've typically seen `ExUnitProperties` tests use [`property/3`](https://hexdocs.pm/stream_data/ExUnitProperties.html#property/3) instead of `ExUnit.Case.test/3`. What is the effect of using `test` instead?

I'd probably go with `property/3`, and use a helper function to generate a user.

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,68 @@
+  describe "calculate_votes/1" do
+    setup do
+      points_map = StreamData.fixed_map(%{points: integer(1..5)})
+      metas_map  = StreamData.fixed_map(%{
```
`use ExUnitProperties` will import `StreamData` functions. We can omit the `StreamData` prefix here.

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,68 @@
+      metas_map  = StreamData.fixed_map(%{
+        metas: list_of(points_map, length: 1),
+      })
+      user = nonempty(map_of(string(:printable), metas_map))
```
This is only a minor preference, but I'd suggest using `:alphanumeric` here. It doesn't generate quite the same range of strings as `:printable` will, but the output is easier to read. I'm trading off some robustness for improved readability. If you wanted to keep `:printable`, though, I can get behind that decision 😄

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,68 @@
+
+        max_vote = users
+        |> Enum.map(fn {_username, %{metas: [%{points: points}]}} -> points end)
+        |> Enum.max()
```
A popular style guide [recommends](https://github.com/christopheradams/elixir_style_guide#pipe-operator) writing pattern-matches with pipes like this:
```ex
max_vote =
  users
  |> Enum.map(fn {_username, %{metas: [%{points: points}]}} -> points end)
  |> Enum.max()
```
Would you be open to that format?

SophieDeBenedetto commented:
Thanks for the feedback @michaelstalker! I addressed your comments and I also refactored the vote calculation module to rely on a `score_card` token that we can just pipe through a series of functions.

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,87 @@
+
+          is_integer(winner) ->
+            winner <= max_vote
+        end
```
This section of code is going always going to pass because it doesn't assert anything. Once you add `assert` to line 52, the test will fail.

SophieDeBenedetto replied:
good catch! fixed.

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,87 @@
+        })
+
+      user = nonempty(map_of(string(:alphanumeric), metas_map))
+      [user: user]
```
I didn't catch this the first time, but `users` would be a better name for this data. It's a map of one or more users.

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,87 @@
+    end
+
+    property "winning value is a list or a integer", %{user: user} do
+      check all(users <- user) do
```
It seems like a convention to omit parentheses when using the `check/1` macro. I'm not sure why.
`¯\_(ツ)_/¯` All the documentation I've seen omits them, though.

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,87 @@
+      check all(users <- user) do
+        {_event, winner} = PointingParty.VoteCalculator.calculate_votes(users)
+
+        if is_list(winner) do
```
Since the entire set of assertions gets wrapped in a conditional, I think it would be better to set up your preconditions in such a way that `winner` will always be a list.

SophieDeBenedetto replied:
So in order to create such a precondition, we have to ensure that the point value for each user shake out such that there is a tie. Given that StreamData is generating a random integer in the given range, is there anyway to control for that? I.e. to ensure that, of the random set of users, the random points for each user are each a different integer?

michaelstalker replied:
The first thing you'll need to do is to change your map size so the map will always have at least two elements. One-element maps will never result in a `winner` value that is a list.

After that, there are few strategies you could employ:
* Use `filter` to discard all maps whose users all have the same point values
* Use `bind` or `bind_filter` to ensure that the second user's point value is different than the first user's point value. I'm not 100% sure how to do this, since I've never used `bind` before. I'm also not 100% sure this would work.
* Generate a one-user map. That user will have a point value in `[0, 1, 3, 5]`. Then, generate another one-user map. When you do this, your generator would remove the first user's point value from the list of possible point values for the second user. Then, you'd generate a map of zero or more users with a random point value. Finally, you'd join all the maps.

I'm fine leaving the conditional for now, since it won't be trivial to craft a generator that guarantees the results we want. I'd like to see if we can make a generator like that prior to the conference, though.

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,87 @@
+            calculated_votes
+            |> Enum.sort_by(&elem(&1, 1))
+            |> Enum.take(2)
+            |> Enum.map(&elem(&1, 0))
```
The code from lines 69-80 is the same as the application code. That's generally not ideal because it won't catch bugs in the original algorithm. Is there a different way we can express this code in the test?

SophieDeBenedetto replied:
Yea I was definitely wondering this myself! I took a stab at finding a _slightly_ different way of expressing it, but there is at least one line that is still the same as application code. Let me know what you think.

michaelstalker commented on test/pointing_party_web/controllers/page_controller_test.exs:
```diff
@@ -4,9 +4,11 @@ defmodule PointingPartyWeb.PageControllerTest do
+    setup %{conn: conn} do
+      conn =
+        conn
+        |> Plug.Conn.assign(:username, @username)
```
I generally try to avoid using a pipe when there's only one function. (Chrisopher Adams's style guide also [recommends](https://github.com/christopheradams/elixir_style_guide#avoid-single-pipelines) avoiding the pipe operator when there's just one `|>`) What do you think about the following code?
```ex
conn = Plug.Conn.assign(conn, :username, @username)
```

I have a similar suggestion for [this code](https://github.com/elixirschool/pointing-party/pull/8/files#diff-51344aeb6586e385c77c3b2642a91c49R26).

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,84 @@
+
+  describe "calculate_votes/1" do
+    setup do
+      points_map = fixed_map(%{points: integer(1..5)})
```
It would probably be better to use this, to match what the user does in the real app more closely:
```ex
points_map = fixed_map(%{points: member_of(0, 1, 3, 5)})
```

michaelstalker replied:
The Elixir code uses `[0, 1, 3, 5]`. The HTML we create in the JavaScript uses `[1, 2, 3, 5]`. We should probably iron that out. Maybe `1..5` is fine for now, as long as we make a to-do to fix it before the workshop.

michaelstalker commented on test/pointing_party/vote_calculator_test.exs:
```diff
@@ -0,0 +1,84 @@
+    end
+
+    property "winning value is a list or a integer", %{users: users} do
+      check all users <- users do
```
:100: for renaming the property parameter from `user` to `users. I discourage people from reassigning values to parameters. (Martin Fowler used to have a refactoring just for this, but reassigning to parameters is now a general case of what he calls ["Split Variable."](https://refactoring.com/catalog/splitVariable.html)) It can create some confusion about what exactly a variable refers to. What would you think about this?
```ex
check all user_map <- users do
```
